### PR TITLE
MCR-2175 resolve thumbnails in IIIF-API

### PR DESCRIPTION
**Pull request by rsteph-de**
[Link to jira](https://mycore.atlassian.net/browse/MCR-2175).

### CODE REVIEW

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java  5
- mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159&bid=17791799)

File: mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java [hunk 1]
```diff
@@ -49,14 +54,38 @@
...
+     *                      return the original size and centered parameter will be
+     *                      ignored
+     * @param pdfFile       - the file from which the thumbnail will be taken
+     * @param centered      - if true, a square (thumbnail with same width and
+     *                      height) will be returned
+     * @return a BufferedImage as thumbnail
+     * 
+     * @throws IOException
+     */
+    public static BufferedImage getThumbnail(int thumbnailSize, Path pdfFile, boolean centered) throws IOException {
```
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java:80]: Codacy Issue found: [The method 'getThumbnail(int, Path, boolean)' has an NPath complexity of 417](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java  2
- mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159&bid=17791799)

File: mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java [hunk 2]
```diff
@@ -49,14 +54,39 @@
...
+     * @param thumbnailSize - the size: size = max(width, height) 
+     *                        a size &lt; 0 will return the original size and centered parameter will be ignored
+     * @param pdfFile       - the file from which the thumbnail will be taken
+     * @param centered      - if true, a square (thumbnail with same width and
+     *                      height) will be returned
+     * @return a BufferedImage as thumbnail
+     * 
+     * @throws IOException
+     */
+    public static BufferedImage getThumbnail(int thumbnailSize, Path pdfFile, boolean centered) throws IOException {
```
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java:81]: Codacy Issue found: [The method 'getThumbnail(int, Path, boolean)' has an NPath complexity of 417](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 9e80c44d0ead9ace1965a4f393f96f40c89d9bff! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/5452-pull_request-955

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 7a543da7c617d6f514feabdb8c6319d493f53751! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/5456-pull_request-955

**Comment by sebhofmann** [comment]
What is wrong with this PR:
- extra code for tiling is added, which is not using the iview job api.   
- iview iiif impl now decides what is the thumbnail, why not make a extra impl. for that decision.
- we generate iview files for pdf pages, seems like a little bit overkill. 

We should keep things simple in mycore-iview. 
In my opinion the pdf stuff belongs to mycore-media.
We need one IIIF impl. that does PDF, one impl that does iview and one which just delegates to the others for a mycore object.

**Comment by rsteph-de** [comment]
>     * extra code for tiling is added, which is not using the iview job api.

I was not able to get this working ...
Iview Job API needs a physical image file - you cannot pass the thumbnail as BufferedImage, ...
since all metadata for the job needs to be stored within database. This is limited and e.g. I did not find an easy solution to cleanup afterwards.

>     * iview iiif impl now decides what is the thumbnail, why not make a extra impl. for that decision.

There is already an interface MCRTileFileProvider
which is implemented by iview and can be reimplemented in other applications.
I wanted to keep things "together" and added a 2nd method for thumbnail retrieval.
For thumbnail generation it is possible to implement your own eventhandler if needed.

>     * we generate iview files for pdf pages, seems like a little bit overkill.

I did not want to do it in the first place, but doing it differently whould need a lot more code changes in the IIIF implementation which works on tiles. On the other side this allows us to handle larger thumbnails - think of them more as "title page" images , which can also serve as thumbnail ;-) 

> We should keep things simple in mycore-iview.

Agreed

> In my opinion the pdf stuff belongs to mycore-media.

I thought myore-media primarily contains code for handling audio and video, which may not be used in all repositories. This may lead to other dependencies in code and configuration. Therefore I did not want to mix that with PDF handling.
I tried to avoid a new dependency from mycore-iiif to mycore-media. And most of the code for PDF thumbnail generation was already part of mycore-iview2.


> We need one IIIF impl. that does PDF, one impl that does iview and one which just delegates to  the others for a mycore object.

This is a bigger task that should not be handled in this PR, but separately.
When implementing this, we might consider doing some things differently and do further optimization.

File: mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java (hunk 2 above)
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java:81]: Codacy Issue found: [The method 'getThumbnail(int, Path, boolean)' has an NPath complexity of 417](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java  2
- mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159&bid=17791799)

**Comment by sebhofmann** [comment]
> Iview Job API needs a physical image file - you cannot pass the thumbnail as BufferedImage,

You already create a temp png file.

 > I thought myore-media primarily contains code for handling audio and video, which may not be used in all repositories. This may lead to other dependencies in code and configuration. Therefore I did not want to mix that with PDF handling.

Then we should consider a mycore-pdf component.

> There is already an interface MCRTileFileProvider
> which is implemented by iview and can be reimplemented in other applications.
> I wanted to keep things "together" and added a 2nd method for thumbnail retrieval.
> For thumbnail generation it is possible to implement your own eventhandler if needed.

Yes, but the problem is much more generic than iview. There could also be Thumbnails for Videos, Audio or any other file type. MCRTileFileProvider is only there to tell mycore where to get the .iview files.

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java  5
- mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java  2
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159&bid=17791799)

File: mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java (hunk 2 above)
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java:81]: Codacy Issue found: [The method 'getThumbnail(int, Path, boolean)' has an NPath complexity of 417](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java  7
- mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java  5
- mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java  2
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159&bid=17791799)

File: mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    public MCRPDFThumbnailJobAction() {
+    }
+
+    public MCRPDFThumbnailJobAction(MCRJob job) {
+        super(job);
+    }
+
+    public static final String DERIVATE_PARAMETER = "derivate";
+
+    private static Logger LOGGER = LogManager.getLogger();
```
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java:35]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

File: mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+public class MCRPDFThumbnailJobAction extends MCRJobAction {
+
+    public MCRPDFThumbnailJobAction() {
+    }
+
+    public MCRPDFThumbnailJobAction(MCRJob job) {
+        super(job);
+    }
+
+    public static final String DERIVATE_PARAMETER = "derivate";
```
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java:33]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

File: mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+                // used OpenOption.DELETE_ON_CLOSE
+                // Files.deleteIfExists(pImg);
+            } catch (IOException e) {
+                LOGGER.error("Error creating thumbnail for PDF", e);
+            }
+        }
+    }
+
+    @Override
+    public void rollback() {
```
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java:79]: Codacy Issue found: [Document empty method body](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

File: mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+    private static final MCRJobQueue PDF_THUMBNAIL_JOB_QUEUE = initializeJobQueue();
+
+
+    private static MCRJobQueue initializeJobQueue() {
+        LOGGER.info("Initializing jobQueue for PDF Thumbnail generation!");
+        return MCRJobQueue.getInstance(MCRPDFThumbnailJobAction.class);
+    }
+
+    private static List<String> derivateTypesForContent = MCRConfiguration2
```
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java:62]: Codacy Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

File: mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java (hunk 2 above)
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java:81]: Codacy Issue found: [The method 'getThumbnail(int, Path, boolean)' has an NPath complexity of 417](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           

Complexity increasing per file
==============================
- mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java  5
- mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java  2
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java  3
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159&bid=17791799)

File: mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
```
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java:79]: Codacy Issue found: [Document empty method body](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

File: mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java [hunk 8]
```diff
@@ -30,33 +30,63 @@
...
+     * @param thumbnailSize - the size: size = max(width, height) 
+     *                        a size &lt; 0 will return the original size and centered parameter will be ignored
+     * @param pdfFile       - the file from which the thumbnail will be taken
+     * @param centered      - if true, a square (thumbnail with same width and
+     *                      height) will be returned
+     * @return a BufferedImage as thumbnail
+     *
+     * @throws IOException
+     */
+    public static BufferedImage getThumbnail(int thumbnailSize, Path pdfFile, boolean centered) throws IOException {
```
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java:81]: Codacy Issue found: [The method 'getThumbnail(int, Path, boolean)' has an NPath complexity of 417](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 48228a67ef88deeff4165951f44a53f10b9deedd! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/5499-pull_request-955

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java  3
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java  7
- mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java  5
- mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159&bid=17791799)

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 21bdc4f9887cfbeb0dcdde84ddd884b4ed47020b! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/5501-pull_request-955

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java  2
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java  3
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java  7
- mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159&bid=17791799)

File: mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java (hunk 8 above)
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java:81]: Codacy Issue found: [The method 'getThumbnail(int, Path, boolean)' has an NPath complexity of 417](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on e14aaa80585f2e97ee8aa4ec628161eda6a0b749! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/5506-pull_request-955

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java  7
- mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java  2
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java  3
- mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159&bid=17791799)

File: mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java [hunk 9]
```diff
@@ -30,60 +30,127 @@
...
-                bg.setRenderingHint(RenderingHints.KEY_INTERPOLATION, RenderingHints.VALUE_INTERPOLATION_BICUBIC);
-                int x = centered ? (thumbnailSize - newWidth) / 2 : 0;
-                int y = centered ? (thumbnailSize - newHeight) / 2 : 0;
-                if (x != 0 && y != 0) {
-                    LOGGER.warn("Writing at position {},{}", x, y);
+            return scalePage(thumbnailSize, centered, level1Image);
+        }
+    }
+
+    private static BufferedImage scalePage(int thumbnailSize, boolean centered, BufferedImage level1Image) {
```
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java:94]: Codacy Issue found: [The method 'scalePage(int, boolean, BufferedImage)' has an NPath complexity of 208](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

File: mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+                mcrDer = MCRMetadataManager.retrieveMCRDerivate(mcrID);
+                return Optional.of(new MCRTileInfo(mcrDer.getId().toString(),
+                        mcrDer.getDerivate().getInternals().getMainDoc(), null));
+            } else {
+                MCRObject mcrObj = MCRMetadataManager.retrieveMCRObject(mcrID);
+                for (String derivateType : derivateTypesForThumbnails) {
+                    for (MCRMetaEnrichedLinkID derLink : mcrObj.getStructure().getDerivates()) {
+                        Optional<String> derType = derLink.getContentList().stream()
+                                .filter(x -> (x instanceof Element))
+                                .map(x -> (Element) x)
```
Comment (yagee-de) [inline mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java]: How about:
```
.filter(Element.class::isInstance)
.map(Element.class::cast)
```

File: mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+public class MCRDefaultThumbnailTileInfoProvider implements MCRThumbnailTileInfoProvider {
+
+    private static List<String> derivateTypesForThumbnails = MCRConfiguration2
+            .getOrThrow("MCRIIIFImage.Iview.Thumbnail.Derivate.Types", MCRConfiguration2::splitValue)
+            .collect(Collectors.toList());
+
+    @Override
+    public Optional<MCRTileInfo> getThumbnailFileInfo(String id) {
+        if (MCRObjectID.isValid(id)) {
```
Comment (yagee-de) [inline mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java]: invert check and return early

File: mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            MCRDerivate mcrDer = null;
+            if (mcrID.getTypeId().equals("derivate")) {
+                mcrDer = MCRMetadataManager.retrieveMCRDerivate(mcrID);
+                return Optional.of(new MCRTileInfo(mcrDer.getId().toString(),
+                        mcrDer.getDerivate().getInternals().getMainDoc(), null));
+            } else {
+                MCRObject mcrObj = MCRMetadataManager.retrieveMCRObject(mcrID);
+                for (String derivateType : derivateTypesForThumbnails) {
+                    for (MCRMetaEnrichedLinkID derLink : mcrObj.getStructure().getDerivates()) {
+                        Optional<String> derType = derLink.getContentList().stream()
```
Comment (yagee-de) [inline mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java]: Do not use streams here if you know where to look for. Use JDOM methods directly and your code is clearly better understandable.

File: mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+                                .filter(x -> (x instanceof Element))
+                                .map(x -> (Element) x)
+                                .filter(e -> e.getName().equals("classification"))
+                                .map(c -> {
+                                    return c.getAttributeValue("classid") + ":" + c.getAttributeValue("categid");
+                                })
+                                .filter(c -> c.equals(derivateType))
+                                .findFirst();
+                        if (derType.isPresent()) {
+                            Optional<String> oMainDoc = derLink.getContentList().stream()
```
Comment (yagee-de) [inline mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java]: Please use JDOM API to get to main. If in doubt use an XPath.

File: mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+                    MCRImage mcrImage = MCRImage.getInstance(pImg, tileInfo.getDerivate(), tileInfo.getImagePath());
+                    mcrImage.setTileDir(MCRIView2Tools.getTileDir());
+                    mcrImage.tile();
+                } finally {
+                    Files.deleteIfExists(pImg);
+                }
+                // RS: throws java.nio.file.AccessDeniedException, when immediately called after
+                // tile()
+                // used OpenOption.DELETE_ON_CLOSE
+                // Files.deleteIfExists(pImg);
```
Comment (yagee-de) [inline mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java]: What about this comment?

File: mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRThumbnailTileInfoProvider.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MyCoRe.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.mycore.iview2.backend;
+
+import java.util.Optional;
+
+public interface MCRThumbnailTileInfoProvider {
```
Comment (yagee-de) [inline mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRThumbnailTileInfoProvider.java:26]: There is literately not a single line of JavaDocs for this interface or its method!

File: mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+public class MCRThumbnailForPdfEventHandler extends MCREventHandlerBase {
+
+    public static final MCRDefaultTileFileProvider TILE_FILE_PROVIDER = new MCRDefaultTileFileProvider();
+
+    private static final Logger LOGGER = LogManager.getLogger(MCRThumbnailForPdfEventHandler.class);
+
+    private static final MCRJobQueue PDF_THUMBNAIL_JOB_QUEUE = initializeJobQueue();
+
+    private static final List<String> DERIVATE_TYPES_FOR_CONTENT = MCRConfiguration2
+        .getOrThrow("MCRIIIFImage.Iview.ThumbnailForPdfEventHandler.Derivate.Types",
```
Comment (yagee-de) [inline mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java]: Does not seem to be a MCR.* property, why?

File: mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+import org.mycore.datamodel.metadata.MCRDerivate;
+import org.mycore.datamodel.metadata.MCRMetaEnrichedLinkID;
+import org.mycore.datamodel.metadata.MCRMetadataManager;
+import org.mycore.datamodel.metadata.MCRObject;
+import org.mycore.datamodel.metadata.MCRObjectID;
+
+public class MCRDefaultThumbnailTileInfoProvider implements MCRThumbnailTileInfoProvider {
+
+    private static List<String> derivateTypesForThumbnails = MCRConfiguration2
+            .getOrThrow("MCRIIIFImage.Iview.Thumbnail.Derivate.Types", MCRConfiguration2::splitValue)
```
Comment (yagee-de) [inline mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java]: Why not use MCR.* as property?

File: mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java [hunk 18]
```diff
@@ -76,38 +75,52 @@
...
         super(implName);
         Map<String, String> properties = getProperties();
 
         String tileFileProviderClassName = properties.get(TILE_FILE_PROVIDER_PROPERTY);
-
         if (tileFileProviderClassName == null) {
-            tileFileProviderClassName = MCRDefaultTileFileProvider.class.getName();
+            tileFileProvider = new MCRDefaultTileFileProvider();
+        } else {
+            Optional<MCRTileFileProvider> optTFP = MCRConfiguration2.getInstanceOf(tileFileProviderClassName);
```
Comment (yagee-de) [inline mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java]: getInstanceOf() uses a property name and not a className as parameter. You might want to use instantiateClass() instead.

File: mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java [hunk 19]
```diff
@@ -76,38 +75,52 @@
...
-            throw new MCRConfigurationException("Configurated class (" + TILE_FILE_PROVIDER_PROPERTY
-                + ") needs a default constructor: " + tileFileProviderClassName);
-        } catch (IllegalAccessException | InstantiationException | InvocationTargetException e) {
-            throw new MCRException(e);
+        String thumbnailTileInfoProviderClassName = properties.get(THUMBNAIL_TILE_INFO_PROVIDER_PROPERTY);
+        if (thumbnailTileInfoProviderClassName == null) {
+            thumbnailTileInfoProvider = new MCRDefaultThumbnailTileInfoProvider();
+        } else {
+            Optional<MCRThumbnailTileInfoProvider> optTTFIP = MCRConfiguration2
+                    .getInstanceOf(thumbnailTileInfoProviderClassName);
```
Comment (yagee-de) [inline mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java]: Same here use instantiateClass()

**Review by yagee-de — CHANGES_REQUESTED** [review]

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java  5
- mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java  2
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java  3
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159&bid=17791799)

File: mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java (hunk 9 above)
Comment (adlerre) [inline mycore-iview2/src/main/java/org/mycore/iview2/frontend/MCRPDFTools.java:94]: Codacy Issue found: [The method 'scalePage(int, boolean, BufferedImage)' has an NPath complexity of 208](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159)

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 75f78cbeec0768ad42dbf67161d8a3823ffbd7c4! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/5513-pull_request-955

**Comment by adlerre** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- mycore-iview2/src/main/java/org/mycore/iview2/iiif/MCRIVIEWIIIFImageImpl.java  2
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRDefaultThumbnailTileInfoProvider.java  6
- mycore-iview2/src/main/java/org/mycore/iview2/backend/MCRPDFThumbnailJobAction.java  3
- mycore-iview2/src/main/java/org/mycore/iview2/events/MCRThumbnailForPdfEventHandler.java  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/MyCoRe/mycore/pullRequest?prid=5425159&bid=17791799)

**Pull request merged**